devito/types/sparse.py

```diff
@@ -451,35 +730,11 @@ class SparseFunction(AbstractSparseFunction):
     def __init_finalize__(self, *args, **kwargs):
         super(SparseFunction, self).__init_finalize__(*args, **kwargs)
         self.interpolator = LinearInterpolator(self)
+
         # Set up sparse point coordinates
         coordinates = kwargs.get('coordinates', kwargs.get('coordinates_data'))
-        if isinstance(coordinates, Function):
-            self._coordinates = coordinates
-        else:
-            dimensions = (self.indices[self._sparse_position], Dimension(name='d'))
-            # Only retain the local data region
-            if coordinates is not None:
-                coordinates = np.array(coordinates)
-            self._coordinates = SubFunction(
-                name='%s_coords' % self.name, parent=self, dtype=self.dtype,
-                dimensions=dimensions, shape=(self.npoint, self.grid.dim),
-                space_order=0, initializer=coordinates, alias=self.alias,
-                distributor=self._distributor
-            )
-            if self.npoint == 0:
-                # This is a corner case -- we might get here, for example, when
-                # running with MPI and some processes get 0-size arrays after
-                # domain decomposition. We "touch" the data anyway to avoid the
-                # case ``self._data is None``
-                self.coordinates.data
-
-    def __distributor_setup__(self, **kwargs):
-        """
-        A `SparseDistributor` handles the SparseFunction decomposition based on
-        physical ownership, and allows to convert between global and local indices.
-        """
-        return SparseDistributor(kwargs['npoint'], self._sparse_dim,
-                                 kwargs['grid'].distributor)
+        self._coordinates = self.__subfunc_setup__(coordinates, 'coords')
+        self._dist_origin = {self._coordinates: self.grid.origin_offset}
```

Comment:
(cached) property ?

Reply:
prod

Reply:
up

Reply:
It is a property in the super class yes

Reply:
What I meant to say is that `_dist_origin` should be a (cached?) property, not attached to self at `__init__` time

Reply:
It depends on the input in Precomputed and has to be set as __init__

Reply:
What's the failing test if I turn it into a property?

the `_dist_origin` dictionary maps two objects that belong to `self`. Since `self` and all its attributes are immutable (clearly aside from the .data), I struggle to see why this should be a problem

Reply:
ah, OK.

But if the Precomputed*'s constructor overrides this, then I think you might rather want to use a `__setup_origin__` method. 

Anyway, nitpicking for another day